various-receipes/component-isolation/test/isolation.test.js

```diff
@@ -116,4 +117,83 @@ describe('/api', () => {
       });
     });
   });
+
+  test('When users service doesn\'t reply within 2 seconds, return 503', async () => {
+    //Arrange
+    nock.cleanAll();
+    nock('http://localhost/user/')
+      .get('/1')
+      .delay(3000)
+      .reply(200);
+    nock('https://mailer.com')
+      .post('/send')
+      .reply(202);
+    const orderToAdd = {
+      userId: 1,
+      productId: 2,
+      mode: 'approved',
+    };
+
+    //Act
+    const response = await request(expressApp).post('/order').send(orderToAdd);
+    
+    //Assert
+    expect(response.status).toBe(503);
+  });
+
+  test('When users service replies with 503, should retry the request', async () => {
+    //Arrange
+    nock.cleanAll();
+    const firstRequest = nock('http://localhost/user/')
+      .get('/1')
+      .reply(503, undefined, { 'Retry-After': 100 });
+    const secondRequest = nock('http://localhost/user/')
+      .get('/1')
+      .reply(200);
+    nock('https://mailer.com')
+      .post('/send')
+      .reply(202);
+    const orderToAdd = {
+      userId: 1,
+      productId: 2,
+      mode: 'approved',
+    };
+
+    //Act
+    const response = await request(expressApp).post('/order').send(orderToAdd);
+
+    //Assert
+    expect(response.status).toBe(200);
+    expect(firstRequest.isDone()).toBeTruthy();
+    expect(secondRequest.isDone()).toBeTruthy();
+  });
+
+  test.only('When request is more than 3 seconds, should fire a metric', async () => {
+    //Assert
+    const clock = sinon.useFakeTimers();
+    
+    sinon
+      .stub(OrderRepository.prototype, 'addOrder')
+      .callsFake((...args) => {
+        clock.tick(5000);
+        return OrderRepository.prototype.addOrder.wrappedMethod(...args)
+      });
+    
+    const orderToAdd = {
+      userId: 1,
+      productId: 2,
+      mode: 'approved',
+    };
+
+    const metricsExporterDouble = sinon.stub(metricsExporter, 'fireMetric');
+
+    //Act
+    await request(expressApp).post('/order').send(orderToAdd);
+    
+    //Assert
+    expect(metricsExporterDouble.calledWith('too slow response')).toBe(true);
+
+    //Clean
+    clock.uninstall();
+  });
```

Comment:
Consider also testing that:

-  When a collaborator (users service) has a delay within the timeout range, the order is added successfully
-  When a collaborator is not answering at all (timeout), then a metric is fired

Reply:
> When a collaborator (users service) has a delay within the timeout range, the order is added successfully

What are we checking in this scenario?

> When a collaborator is not answering at all (timeout), then a metric is fired

We have a test with a timeout. there is something special in this `then`?